File: jlpt/native_android/app/src/main/java/com/jlpt/android/data/repository/AuthRepositoryImpl.kt
package com.jlpt.android.data.repository

import com.jlpt.android.core.datastore.TokenDataStore
import com.jlpt.android.data.mapper.toDomain
import com.jlpt.android.data.remote.api.AuthApi
import com.jlpt.android.data.remote.dto.LoginRequestDto
import com.jlpt.android.data.remote.dto.RegisterRequestDto
import com.jlpt.android.data.remote.safeApiCall
import com.jlpt.android.domain.model.AuthSession
import com.jlpt.android.domain.model.JlptLevel
import com.jlpt.android.domain.repository.AuthRepository
import com.jlpt.android.domain.util.AppResult
import com.jlpt.android.domain.util.onSuccess
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepositoryImpl @Inject constructor(
    private val api: AuthApi,
    private val tokenDataStore: TokenDataStore,
    private val json: Json
) : AuthRepository {

    override val session: Flow<AuthSession?> = tokenDataStore.session

    override suspend fun login(email: String, password: String): AppResult<AuthSession> =
        safeApiCall(json) { api.login(LoginRequestDto(email, password)).toDomain() }
            .onSuccess { tokenDataStore.save(it) }

    override suspend fun register(
        email: String,
        password: String,
        displayName: String,
        targetLevel: JlptLevel?
    ): AppResult<AuthSession> =
        safeApiCall(json) {
            api.register(RegisterRequestDto(email, password, displayName, targetLevel)).toDomain()
        }.onSuccess { tokenDataStore.save(it) }

    override suspend fun logout() {
        tokenDataStore.clear()
    }
}

File: jlpt/native_ios/Sources/Data/Repositories/AuthRepositoryImpl.swift
import Foundation

final class AuthRepositoryImpl: AuthRepository {
    private let api: AuthAPI
    private let tokenStore: TokenStore

    init(api: AuthAPI, tokenStore: TokenStore) {
        self.api = api
        self.tokenStore = tokenStore
    }

    func sessionUpdates() -> AsyncStream<AuthSession?> {
        tokenStore.sessionUpdates()
    }

    func login(email: String, password: String) async -> AppResult<AuthSession> {
        let result = await safeCall { try await self.api.login(LoginRequestDTO(email: email, password: password)).toDomain() }
        if case .success(let session) = result {
            await tokenStore.save(session)
        }
        return result
    }

    func register(
        email: String,
        password: String,
        displayName: String,
        targetLevel: JlptLevel?
    ) async -> AppResult<AuthSession> {
        let request = RegisterRequestDTO(email: email, password: password, displayName: displayName, targetLevel: targetLevel)
        let result = await safeCall { try await self.api.register(request).toDomain() }
        if case .success(let session) = result {
            await tokenStore.save(session)
        }
        return result
    }

    func logout() async {
        await tokenStore.clear()
    }
}
